### PR TITLE
Homepage Posts: Rework how image shapes are treated

### PR DESCRIPTION
**Pull request by laurelfulford**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR replaces #503 - though the underlying approach is the same, since it feels ready for more eyes/options, I thought it would be good to start a clean PR with only relevant information.

**Approach**

This PR removes all image sizes from the Homepage Posts block. They are replaced with a combination of:

* CSS, to assist with creating the different image shapes (landscape, portrait, square)
* Using Jetpack Photon to size images on the front-end when available
* Using WordPress’s baked-in ‘Large’ image size in the editor, and as a fallback on the front-end when Jetpack isn’t installed. 

There's a lot of information here to highlight issues, and limited screenshots -- a more complete WP.com test site can be found [here](https://imagesizetest2020.wordpress.com/). There's more testing information at the end of the issue.

----

### Potential issues/noticeable changes

<details>

**<summary>Too-small images treated differently</summary>**

Originally, the block made post thumbnails 100% of the available width, even when they were too small; when they were very small, they were also uncropped -- this is an extreme example, with these smaller images sized up to 1200px wide -- the blue border running around the first image shows the original was too small to generate a cropped version; it doesn't display at the 4:3 aspect ratio that should be used for the 'landscape' shape, and both images look terrible:

image

Now the styles use the image’s original size to crop it, so even very small images are cropped to the correct shape.

They are also not scaled up, so too small images will display at their actual size. This is largely because Varia includes styles to stop the images from being scaled up, and the approach in the block now includes some inline styles, which would be difficult to override.

The above images (with some larger versions) now look like this:

image

On Newspack sites, the images didn't actually scale up when AMP was enabled. So prior to these changes, they will already display smaller; after, the only difference would be that the first image would be cropped: 

image

[Example of "before" with Twenty Twenty](https://cloudup.com/iQSDfxB9qDm/f) (unlike Varia, it doesn’t override the 100% width)
[Example of after with Twenty Twenty](https://cloudup.com/i2X7HISVNxU/f)

</details>

<details>

**<summary>`srcset` attributes have been removed</summary>**

WordPress, by default will serve up responsive versions of images using srcset; these are sized to fit their containing element. 

Because we’re placing the image in a smaller element to crop it, using srcset causes a smaller version of the image — the same width as the container — to be loaded, causing some image quality issues.

I’ve opted to remove `srcset` from the images altogether, but this feels a bit heavy-handed -- I’d appreciate some feedback on this, in case there’s a better way.

</details>

<details>

**<summary>”Large” image as fallback, and as default in the editor</summary>**

The block will use Photon to size images when Jetpack is installed, even when Photon is not enabled (not due to anything calculated, but just because `class_exists( 'Jetpack_PostImages' )` still true when just Jetpack is enabled, and `Jetpack_PostImages::fit_image_url` returns an image). 

When Jetpack is not installed, the block will fall back to the ‘Large’ image size, defined under WP Admin > Settings > Media. 

The editor will always use the ‘Large’ image.

The default size for the 'Large' image is 1024x1024, but it will be scaled up to fit the original image's potential size, so it can have a slightly degraded quality compared to the 1200px wide version originally generated by the block.

I think this is an acceptable loss: the 'Large' image size is larger than most site's content areas, and larger than most sites would typically display this block.

The size is also ediable; for Newspack sites (where the content width is 1200px with the one column wide template), we could switch the 'Large' size to use 1200px x 9999px.

</details>

<details>

**<summary>Changes to ‘Image Behind’ position option</summary>**

As part of this update, I made some changes to the ‘Image Behind’ position styles:

* The ‘Image Shape’ (Landscape, Portrait, Square) option is removed in the editor when the ‘Image Behind’ position in selected
* The full-sized image (either 1200x9999 with Photon, or ‘Large’) is used when ‘Image Behind’ is selected
* The image shape classes (`is-landscape`, `is-portrait`, `is-square`) are no longer applied.

For the most part this doesn’t change anything about how this style functions — it was mostly to avoid writing a bunch of redundant or `:not()`-riddled CSS.

However, because image sizes are no longer cropped, there may be a small quality improvement in some cases, like when the original image is smaller — you can see this in the first couple posts here:

* [Twenty Twenty and Image Behind before](https://cloudup.com/iiyDRbVbQ_9)
* [Twenty Twenty and Image Behind after](https://cloudup.com/iHnTBxXnLmI)

</details>

----

### Newspack-specific issues and notes:

<details>

**<summary>Custom CSS and Newspack</summary>**

 Three publishers include custom CSS for the block images:

* Technode has styles to add rounded corners to the images in the block. With these changes, ‘overflow: hidden’ will need to be added to their styles to keep this working. It won't cause issues, so can be added at any time.
* El Soberano also has Custom CSS that adds rounded corners, but it already has the `overflow: hidden` so it will display the same after these updates.
* Oaklandside uses CSS to update the image crop; it will need to be updated. It can’t be updated until the block is updated, so it will need to be changed there with some care.

</details>

<details>

**<summary>Newspack Theme screenshots</summary>**

Here are a set of screenshots showing how too-small images look with the Newspack theme, before and after the changes.

With the after version ‘no Jetpack’ screenshots, if you zoom in you can also note the slight degradation of image quality on the larger images by looking at the numbers — this is from scaling up a 1024px wide image to 1200px wide. 

* [Newspack landscape images - no AMP, no Jetpack](https://cloudup.com/cxmOSDv50gu)
* [Newspack landscape images - AMP, no Jetpack](https://cloudup.com/c5bjoVCbWqJ)
* [Newspack landscape images - no AMP, with Jetpack](https://cloudup.com/cITpQpWKPcq)
* [Newspack landscape images - with AMP, with Jetpack](https://cloudup.com/c2YCXCSGyyp)
* [Newspack behind images - no AMP, no Jetpack](https://cloudup.com/c8UYfihrwER)
* [Newspack behind images - AMP, no Jetpack](https://cloudup.com/c8jSEawOMC6) 
* [Newspack behind images - no AMP, with Jetpack](https://cloudup.com/c9WhZ3N5ql6)
* [Newspack behind images - with AMP, with Jetpac](https://cloudup.com/c8yUwsEgAfi)k

</details>

----

### WP.com-specific issues and notes

For WP.com, I started by testing Varia (since it's the base for WP.com's new themes, and is one theme that has Homepage Posts Block-specific styles), and Twenty Twenty (as a modern, but generic, representation for WP.com themes). 

<details>

**<summary>Varia</summary>**

#### Front-end

Because Varia already removes the 100% widh from the block images, the only difference with smaller images is that the ones that were too small to crop before are now cropped -- these are on the very small size (eg. less than 200px wide). 

This is an edge case, so it’s unlikely to cause issues, but it can be fixed by selecting ‘uncropped’. 

Here are a collection of screenshots with before and after and Varia:

* [Varia landscape images](https://cloudup.com/c0EGm8AsUNB)
* [Varia behind images](https://cloudup.com/cRuQwIscBz6)
* [Varia landscape images - editor](https://cloudup.com/cRrIuXWJiAS) 
* [Varia behind images - editor](https://cloudup.com/csPLsHKZtLZ) 

#### Editor

Varia also has this style in the editor styles:

`.editor-styles-wrapper .wp-block-a8c-blog-posts .post-thumbnail img {width: auto;}`

... that causes the images to be cropped off-center in the editor (you can tell because the blue border on the image appears on the left, and the size is not centred):

image

It's fairly minor (since it would be less noticeable with less extreme image sizes), and removing that style from the theme should fix it.

</details>

<details>

**<summary>Twenty Twenty</summary>**

I tested Twenty Twenty as a representative for themes that don’t have Homepage Posts Block-specific styles.

Outside of too-small images not getting scaled up anymore, Twenty Twenty does not have noticeable issues — here are a set of screenshots, before and after, for comparison: 

* Twenty Twenty landscape images: https://cloudup.com/c_HEM7HEA0N
* Twenty Twenty behind images: https://cloudup.com/c_8iSuOMfmw
* Twenty Twenty landscape images - editor: https://cloudup.com/cPMhXYjP3ST
* Twenty Twenty behind images - editor: https://cloudup.com/c0XZbD6_DS9

</details>

<details>

**<summary>Full-site editing</summary>**

The starter pages in Full-site editing include seven layouts for blogs. General notes:

* Only the landscape and ‘no crop’ shapes are used for images.
* All of these layouts use full-width, so they’re variable in width.
* Outside of too-small images, there are very few differences, especially with Varia. Before and after screenshots for Varia and Twenty Twenty can be found here:

* [Varia and FSE Blog Templates](https://cloudup.com/cbgqxuKQyca)
* [Twenty Twenty and FSE Blog Templates](https://cloudup.com/cI3E_mN2_rb)

</details>

<details>

**<summary>Older themes</summary>**

There are already known issues with the Homepage Posts blocks and some older themes, because the block uses generic theme classes, like `.post-thumbnail` and `.entry-title`, and is wrongly picking up styles (or some JavaScript!) from the themes.

Unfortunately, there's no way for us to fix most of these in the block -- the styles in the themes themselves need to be made more specific so the block doesn't pick them up. 

I do think these changes need more testing with older themes, but there is a chance they could introduce new issues that would be `wontfix`es on the theme level.

</details>

----

### How to test the changes in this Pull Request:

The visual part of this PR is going to be a pain to test, because there are so many variables -- the theme, whether you're on WP.com or self-hosted, Jetpack or AMP, and the original image sizes-- to consider. So rather than the usual steps, I tried putting together some helpful tips and tools:

**Testing tools**
* On WP.com, you can sandbox and use this test site: https://imagesizetest2020.wordpress.com/. It's been set up to have many images of different aspect ratios (16:9, 4:3, 3:2, 1:1, 2:3, 3:4, 9:16) and sizes (widths ranging from 200px - 2000px). Each image has its original aspect ratio and size on it, and a blue border around it so you can see if/when it's cropped and how (top/bottom, and whether the crop is centred). The site also has pages with the block set to different image shapes (landscape, portrait, square, uncropped), in different placements (above, beside, and behind, in 1, 2 and 3 columns), as well as the FSE templates at the bottom. 
* There's an XML export of the above site [here](https://cloudup.com/ix4ATaITm1e) for your own WP.com sandbox, or one [here](https://cloudup.com/i1SPKWYhdZw) that has had the block name swapped for self-hosted sites. In both cases, you unfortunately need to do a bit of manual adjustment of the pages:
     * The import will add a new page for each test case ([landscape image above content](https://imagesizetest2020.wordpress.com/landscape-1-column/), [portrait image beside content](https://imagesizetest2020.wordpress.com/portrait-beside-1-column/), etc.). Each of those pages will load posts with the different aspect ratio images, sorted by category. When you import the content, the category IDs in the blocks will not match anything on your site. Each page needs to be edited, and the category manually selected to match the header on the block:
image
      * You could probably also use a find-and-replace plugin to do the above, if you wanted to edit them all. I think there are a lot of superfluous test pages in that data, though, and you could probably get away with the one-column above and behind version of each image shape, and not necessarily the beside or multi-column views.
* To test on a WP.com sandbox, on top of following the steps here [here](https://github.com/Automattic/wp-calypso/tree/master/apps/full-site-editing#build-system), you need to sandbox the REST API to make sure you’re pulling the new images for the editor.
* To test on self-hosted, you'll need a "real" (not localhost) URL to test Photon. 

In all cases, the before and after for the different configurations need to be tested, in the editor and on the front end. 

On self-hosted sites, testing with Jetpack, with AMP, with both and with neither will also be necessary.

Lastly, testing in different browsers (including mobile, Safari and IE11) is also important.

From the code perspective, I went back and forth on the approach so many times, I'd appreciate any thoughts and suggestions, as well as improvements -- it's very likely I made this more complicated than I needed to. I'm also happy to answer questions! 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by laurelfulford** [comment]
Thanks @dkoo! added a bit more information about using the XML export of the https://imagesizetest2020.wordpress.com site above. Unfortunately, the blocks in the test content use categories to sort the different image aspect ratios, and the category IDs in the XML file won't match the IDs the categories are given when they are imported into a new site.

Hopefully the update makes sense, but if not, please let me know!

File: src/blocks/homepage-articles/edit.js [hunk 1]
```diff
@@ -120,6 +120,25 @@ class Edit extends Component {
 				minHeight / 5 + 'vh',
 		};
 
+		let imgMaxWidth;
```
[the remark this answers is not in the document]
Comment (laurelfulford) [reply src/blocks/homepage-articles/edit.js]: D'oh -- thanks @dkoo! This was a pretty spectacular test case to miss 😄 I've added these checks in cc17f66.

File: class-newspack-blocks.php [hunk 2]
```diff
@@ -208,104 +200,155 @@ public static function is_amp() {
...
-					400,
-				),
-				'tiny'   => array(
-					200,
-					200,
-				),
-			),
+	public static function get_thumbnail_info( $post_id ) {
+		// Get the image.
+		$thumb_id = get_post_thumbnail_id( $post_id );
```
[the remark this answers is not in the document]
Comment (laurelfulford) [reply class-newspack-blocks.php:210]: Good call -- this has been added in 24796b3!

File: class-newspack-blocks.php [hunk 3]
```diff
@@ -208,104 +200,155 @@ public static function is_amp() {
...
+	/**
+	 * Returns cropped image sizes.
+	 *
+	 * @param string $post_id Post ID.
+	 * @return array Array of the width and height of the resized image for each shape.
+	 */
+	public static function get_cropped_size( $post_id ) {
+
+		// Gets the image info.
+		$img_info = self::get_thumbnail_info( $post_id );
```
[the remark this answers is not in the document]
Comment (laurelfulford) [reply class-newspack-blocks.php:240]: Thanks -- added in 24796b3!

File: class-newspack-blocks.php [hunk 4]
```diff
@@ -208,104 +200,155 @@ public static function is_amp() {
...
+
+	/**
+	 * Returns max-width of featured image based on original size and selected shape.
+	 *
+	 * @param string $post_id Post ID.
+	 * @param string $display_shape Shape of image.
+	 * @return array
+	 */
+	public static function get_maxwidth( $post_id, $shape ) {
+		$image   = wp_get_attachment_metadata( get_post_thumbnail_id( $post_id ) );
```
[the remark this answers is not in the document]
Comment (laurelfulford) [reply class-newspack-blocks.php:356]: And added in 24796b3! 😄

**Comment by dkoo** [comment]
As discussed further with Laurel, I need to do more before/after, AMP/no-AMP, and Jetpack/no-Jetpack testing on this using the test content from the Image Size Test site dump. Will post updates from testing here.

File: class-newspack-blocks.php [hunk 5]
```diff
@@ -208,104 +200,166 @@ public static function is_amp() {
...
+			$cropped = self::get_cropped_size( $post_id );
 
-		add_image_size( 'newspack-article-block-landscape-tiny', 200, 150, true );
-		add_image_size( 'newspack-article-block-portrait-tiny', 150, 200, true );
-		add_image_size( 'newspack-article-block-square-tiny', 200, 200, true );
+			$photon['width']  = $cropped[ $shape ]['width'];
+			$photon['height'] = $cropped[ $shape ]['height'];
+
+			// Get the image url.
+			$img_info = self::get_thumbnail_info( $post_id );
```
[the remark this answers is not in the document]
Comment (laurelfulford) [reply class-newspack-blocks.php:391]: Thanks! This one should be fixed in b385d45.

**Comment by laurelfulford** [comment]
Hey @dkoo! I reworked the code that gets the image crop size, to rejig the images so they're always 1200px wide or narrower in 792dc9f -- that was a good call! Let me know if any of that doesn't make sense.

File: class-newspack-blocks.php [hunk 6]
```diff
@@ -208,104 +200,227 @@ public static function is_amp() {
...
+				'width'  => 1200,
+				'height' => 9999,
+			);
 
-		add_image_size( 'newspack-article-block-landscape-medium', 800, 600, true );
-		add_image_size( 'newspack-article-block-portrait-medium', 600, 800, true );
-		add_image_size( 'newspack-article-block-square-medium', 800, 800, true );
+			// Set a cropped height and width if the images are not aligned behind.
+			if ( 'behind' !== $alignment ) {
+				$photon['width']  = $cropped[ $shape ]['width'];
```
Comment (dkoo) [inline class-newspack-blocks.php:386]: @laurelfulford `PHP Notice:  Trying to access array offset on value of type bool in /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-blocks/class-newspack-blocks.php on line 386`. Should probably return early here if `$img_info` is empty.

**Comment by laurelfulford** [comment]
Given the age of this PR I'm going to close it; the 'multiple images generated' issue was solved for sites hosted on the Atomic platform by using Jetpack to generated the crops instead. If we opt to revisit this work, we'll probably need to start fresh!

**Pull request closed without merging**